src/genh5_dataset.cpp

```diff
@@ -16,25 +16,86 @@
 #include "H5Dpublic.h"
 #include "H5Ppublic.h"
 
-GenH5::DataSet::DataSet() = default;
+#include <QDebug>
+
+struct GenH5::DataSet::Impl
+{
+    explicit Impl(hid_t id)
+        : id(id)
+    {}
+
+    /// dataset id
+    IdComponent<H5I_DATASET> id;
+
+    // The hooks to execute before and after read / write
+    std::map<GenH5::HookType, GenH5::Hook> hooks;
+};
+
+GenH5::DataSet::DataSet()
+    : GenH5::DataSet(hid_t())
+{}
 
 GenH5::DataSet::DataSet(hid_t id) :
-    m_id(id)
+    pimpl(std::make_unique<Impl>(id))
+{
+    pimpl->id.inc();
+}
+
+GenH5::DataSet::DataSet(const DataSet & other)
+    : pimpl(std::make_unique<Impl>(*other.pimpl))
+{
+}
+
+GenH5::DataSet::DataSet(DataSet && other)
+    : pimpl(std::move(other.pimpl))
+{
+}
+
+GenH5::DataSet&
+GenH5::DataSet::operator=(const DataSet& other)
 {
-    m_id.inc();
+    GenH5::DataSet tmp(other);
+    swap(tmp);
+    return *this;
 }
 
+GenH5::DataSet&
+GenH5::DataSet::operator=(DataSet&& other)
+{
+    swap(other);
```

Comment:
I am wondering, shouldn't it be like this?
```suggestion
    GenH5::DataSet tmp(std::move(other));
    swap(other);
```

Reply:
I think we both are wrong. In fact a simple

```
swap(std::move(other));
```

should be okay. No need to create a temp object here i think.